### PR TITLE
[Docs] Update about customizing mosaic list

### PR DESCRIPTION
**Pull request by jo66**
## Subject

I am targeting this branch, because it's doc.

### CODE REVIEW

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst [hunk 1]
```diff
@@ -90,7 +91,7 @@ for instance the MediaBundle defines the method as::
 
         public function getObjectMetadata($object)
         {
-            $provider = $this->pool->getProvider($object->getProviderName());
+            $provider = $this->getConfigurationPool()->getContainer()->get($object->getProviderName());
```
Comment (OskarStark) [inline docs/cookbook/recipe_customizing_a_mosaic_list.rst]: why this change? does the previous code not work?

Comment (jo66) [reply docs/cookbook/recipe_customizing_a_mosaic_list.rst]: Yes, the following exception occurs:
`An exception has been thrown during the rendering of a template ("Notice: Undefined property: App\Admin\AuthUserAdmin::$pool").`

Comment (jo66) [reply docs/cookbook/recipe_customizing_a_mosaic_list.rst]: In fact the code works good in BaseMediaAdmin, where it comes from, but not in my own admin. So  I guess I should not change code, but add note to explain how to get provider in general?

Comment (jo66) [reply docs/cookbook/recipe_customizing_a_mosaic_list.rst]: Done

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst [hunk 2]
```diff
@@ -98,6 +99,9 @@ for instance the MediaBundle defines the method as::
         }
     }
 
+.. note::
+    To get the provider from your ``Admin`` class, use this:
+    ``$provider = $this->getConfigurationPool()->getContainer()->get($object->getProviderName());``
```
Comment (core23) [inline docs/cookbook/recipe_customizing_a_mosaic_list.rst]: This chaining of methods is a bad practice. Can you please add a different example with DI

Comment (jo66) [reply docs/cookbook/recipe_customizing_a_mosaic_list.rst]: Done. Thanks for advise.

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst [hunk 3]
```diff
@@ -98,6 +99,69 @@ for instance the MediaBundle defines the method as::
...
+    ``$provider = $this->getConfigurationPool()->getContainer()->get($object->getProviderName());``.
+
+    Another better option is to use dependency injection. For this, first define
+    the ``$pool`` variable and override the constructor:
+
+    .. code-block:: php
+
+        <?php
+
+            ...
```
Comment (OskarStark) [inline docs/cookbook/recipe_customizing_a_mosaic_list.rst]: Please use ‘// ...’ to have valid php code same below

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst [hunk 4]
```diff
@@ -98,6 +99,69 @@ for instance the MediaBundle defines the method as::
...
+                tags:
+                    - name: sonata.admin
+                    manager_type: orm
+                    group: "Content"
+                    label: "Post"
+                arguments:
+                    - ~
+                    - App\Entity\Post
+                    - ~
+                    - '@sonata.media.pool'
```
[the remark this answers is not in the document]
Comment (OskarStark) [reply docs/cookbook/recipe_customizing_a_mosaic_list.rst]: IMO it is more clear like this...

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst [hunk 5]
```diff
@@ -98,6 +99,69 @@ for instance the MediaBundle defines the method as::
...
+                $url = $provider->generatePublicUrl($media, $provider->getFormatName($media, 'admin'));
+
+                return new Metadata($media->getName(), $media->getDescription(), $url);
+            }
+
+    Also, the ``$pool`` variable is not available so you will have to replace
+    ``$provider = $this->pool->getProvider($media->getProviderName());`` by
+    ``$provider = $this->getConfigurationPool()->getContainer()->get($object->getProviderName());``.
+
+    Another better option is to use dependency injection. For this, first define
```
Comment (OskarStark) [inline docs/cookbook/recipe_customizing_a_mosaic_list.rst]: We should only promote the DI way, instead of getting it via the container

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (greg0ire) [reply docs/cookbook/recipe_customizing_a_mosaic_list.rst]: yeah maybe…

[the remark this answers is not in the document]
Comment (jo66) [reply docs/cookbook/recipe_customizing_a_mosaic_list.rst]: Don't know how to do it, I try to follow doc [here](https://symfony.com/doc/current/service_container/autowiring.html) but without success. Maybe you can show me the way?

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst (hunk 3 above)
Comment (jo66) [reply docs/cookbook/recipe_customizing_a_mosaic_list.rst]: Done

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst (hunk 5 above)
Comment (jo66) [reply docs/cookbook/recipe_customizing_a_mosaic_list.rst]: Done

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (jo66) [reply docs/cookbook/recipe_customizing_a_mosaic_list.rst]: When I replace my admin service definition by alias like that:
`Sonata\MediaBundle\Provider\Pool: '@sonata.media.pool'`
I get an error saying that too few arguments are passed to constructor.
When I keep my admin service definition alias seems not needed...

[the remark this answers is not in the document]
Comment (greg0ire) [reply docs/cookbook/recipe_customizing_a_mosaic_list.rst]: Did you specify `autowire: true`?

[the remark this answers is not in the document]
Comment (greg0ire) [reply docs/cookbook/recipe_customizing_a_mosaic_list.rst]: I think we cannot go for autowire until the alias is contributed to the media bundle.

[the remark this answers is not in the document]
Comment (jo66) [reply docs/cookbook/recipe_customizing_a_mosaic_list.rst]: OK, so I resolve

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst [hunk 6]
```diff
@@ -98,6 +99,65 @@ for instance the MediaBundle defines the method as::
...
+
+        <?php
+
+            // ...
+
+            private $pool;
+
+            // ...
+
+            public function __construct($code, $class, $baseControllerName, Pool $pool)
```
[the remark this answers is not in the document]
Comment (jo66) [reply docs/cookbook/recipe_customizing_a_mosaic_list.rst]: So, something to change here?

[the remark this answers is not in the document]
Comment (greg0ire) [reply docs/cookbook/recipe_customizing_a_mosaic_list.rst]: Yes, 2 `string` type hints to add please

[the remark this answers is not in the document]
Comment (jo66) [reply docs/cookbook/recipe_customizing_a_mosaic_list.rst]: Done.

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst [hunk 7]
```diff
@@ -98,6 +99,65 @@ for instance the MediaBundle defines the method as::
...
+            }
+
+    Then add ``'@sonata.media.pool'`` to your service definition:
+
+    .. code-block:: yaml
+
+        services:
+            app.admin.post:
+                class: App\Admin\PostAdmin
+                tags:
```
Comment (OskarStark) [inline docs/cookbook/recipe_customizing_a_mosaic_list.rst]: please move tags below the `arguments` key

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst [hunk 8]
```diff
@@ -98,6 +99,65 @@ for instance the MediaBundle defines the method as::
...
+                    - name: sonata.admin
+                    manager_type: orm
+                    group: "Content"
+                    label: "Post"
+                arguments:
+                    - ~
+                    - App\Entity\Post
+                    - ~
+                    - '@sonata.media.pool'
+                public: true
```
Comment (OskarStark) [inline docs/cookbook/recipe_customizing_a_mosaic_list.rst]: please remove this line

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst [hunk 9]
```diff
@@ -98,6 +99,65 @@ for instance the MediaBundle defines the method as::
...
+                parent::__construct($code, $class, $baseControllerName);
+
+                $this->pool = $pool;
+            }
+
+    Then add ``'@sonata.media.pool'`` to your service definition:
+
+    .. code-block:: yaml
+
+        services:
```
Comment (OskarStark) [inline docs/cookbook/recipe_customizing_a_mosaic_list.rst]:
```suggestion

        # config/services.yaml
        services:
```

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst [hunk 10]
```diff
@@ -98,6 +99,65 @@ for instance the MediaBundle defines the method as::
...
+            }
+
+    You will also have to use dependency injection. For this, first define
+    the ``$pool`` variable and override the constructor:
+
+    .. code-block:: php
+
+        <?php
+
+            // ...
```
Comment (OskarStark) [inline docs/cookbook/recipe_customizing_a_mosaic_list.rst]: please add the use-statemnt of the `Pool` class here

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst [hunk 11]
```diff
@@ -98,6 +99,65 @@ for instance the MediaBundle defines the method as::
...
+
+            private $pool;
+
+            // ...
+
+            public function __construct(string $code, string $class, string $baseControllerName, Pool $pool)
+            {
+                parent::__construct($code, $class, $baseControllerName);
+
+                $this->pool = $pool;
```
Comment (OskarStark) [inline docs/cookbook/recipe_customizing_a_mosaic_list.rst]: its a personal opinion but I would assign the variable first and call the parent one afterwards

WDYT @greg0ire @kunicmarko20 @core23 ?

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst [hunk 12]
```diff
@@ -98,6 +99,65 @@ for instance the MediaBundle defines the method as::
...
+    In your own admin, media is just a field and not the ``$object``. Therefore,
+    the code above must be updated this way:
+
+    .. code-block:: php
+
+        <?php
+
+            // ...
+
+            public function getObjectMetadata($object)
```
Comment (OskarStark) [inline docs/cookbook/recipe_customizing_a_mosaic_list.rst]: the indention is wrong, it needs to be on the same position like the `<?php`-tag. Same for the other php block below

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst [hunk 13]
```diff
@@ -98,6 +99,70 @@ for instance the MediaBundle defines the method as::
...
+        <?php
+
+        // ...
+
+        public function getObjectMetadata($object)
+        {
+            $media = $object->getMediaField();
+
+            $provider = $this->pool->getProvider($media->getProviderName());
+
```
Comment (greg0ire) [inline docs/cookbook/recipe_customizing_a_mosaic_list.rst]:
```suggestion
```

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst (hunk 11 above)
Comment (greg0ire) [reply docs/cookbook/recipe_customizing_a_mosaic_list.rst]: Isn't that exactly what is being done here?

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst [hunk 14]
```diff
@@ -98,6 +99,70 @@ for instance the MediaBundle defines the method as::
...
+            // ...
+
+            public function __construct(string $code, string $class, string $baseControllerName, Pool $pool)
+            {
+                $this->pool = $pool;
+
+                parent::__construct($code, $class, $baseControllerName);
+            }
+
+    Then add ``'@sonata.media.pool'`` to your service definition:
```
Comment (greg0ire) [inline docs/cookbook/recipe_customizing_a_mosaic_list.rst]:
```suggestion
    Then add ``'@sonata.media.pool'`` to your service definition arguments:
```

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst [hunk 15]
```diff
@@ -79,8 +79,9 @@ Block types:
  - ``sonata_mosaic_description``: this block will be always on screen and should represent the entity's name.
 
 
-The ``ObjectMetadata`` object is returned by the related admin class,
-for instance the MediaBundle defines the method as::
+The ``ObjectMetadata`` object is returned by the related admin class, and can be
+used to define which image from entity will be displayed if available. For instance,
+the MediaBundle defines the method as::
```
[the remark this answers is not in the document]
Comment (OskarStark) [reply docs/cookbook/recipe_customizing_a_mosaic_list.rst]: SonataMediaBundle?

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst [hunk 16]
```diff
@@ -98,6 +99,70 @@ for instance the MediaBundle defines the method as::
...
+    In your own admin, media is just a field and not the ``$object``. Therefore,
+    the code above must be updated this way:
+
+    .. code-block:: php
+
+        <?php
+
+        // ...
+
+        public function getObjectMetadata($object)
```
[the remark this answers is not in the document]
Comment (OskarStark) [reply docs/cookbook/recipe_customizing_a_mosaic_list.rst]: But we support PHP 5.6 🤔

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst (hunk 11 above)
Comment (OskarStark) [reply docs/cookbook/recipe_customizing_a_mosaic_list.rst]: Yes now it is done 😊

**Comment by jo66** [comment]
@greg0ire @kunicmarko20 @core23 @OskarStark 
> Please consider contributing the alias for the pool to the media bundle

Sure, but should it be integrated to an existing chapter or should I create new one?

Also, I noticed that I can get provider like that:
`$provider = $this->container->get($media->getProviderName());`
Is it better to use pool or container?

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst (hunk 15 above)
[the remark this answers is not in the document]
Comment (greg0ire) [reply docs/cookbook/recipe_customizing_a_mosaic_list.rst]: Or that, yeah.

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst (hunk 16 above)
[the remark this answers is not in the document]
Comment (greg0ire) [reply docs/cookbook/recipe_customizing_a_mosaic_list.rst]: Docs should assume php 7.3, and this is valid code: https://3v4l.org/srOm8

**Comment by greg0ire** [comment]
> Sure, but should it be integrated to an existing chapter or should I create new one?

It shouldn't be contributed as docs, It should just be a new alias contributed to https://packagist.org/packages/sonata-project/media-bundle

When I register this bundle, there should be an alias pointing from `Sonata\MediaBundle\Provider\Pool` to `'@sonata.media.pool'`

 Is it better to use pool or container?

It is better to use the pool, try never ever to use the container.

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst [hunk 17]
```diff
@@ -79,8 +79,9 @@ Block types:
  - ``sonata_mosaic_description``: this block will be always on screen and should represent the entity's name.
 
 
-The ``ObjectMetadata`` object is returned by the related admin class,
-for instance the MediaBundle defines the method as::
+The ``ObjectMetadata`` object is returned by the related admin class, and can be
+used to define which image from entity will be displayed if available. For instance,
```
[the remark this answers is not in the document]
Comment (jo66) [reply docs/cookbook/recipe_customizing_a_mosaic_list.rst]: Done.

**Comment by jo66** [comment]
@greg0ire Thanks!

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst [hunk 18]
```diff
@@ -98,6 +99,68 @@ for instance the MediaBundle defines the method as::
...
+
+        <?php
+
+        // ...
+
+        use Sonata\MediaBundle\Provider\Pool;
+
+        // ...
+
+            private $pool;
```
Comment (OskarStark) [inline docs/cookbook/recipe_customizing_a_mosaic_list.rst]: Indention is wrong

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst (hunk 15 above)
[the remark this answers is not in the document]
Comment (jo66) [reply docs/cookbook/recipe_customizing_a_mosaic_list.rst]: Updated to SonataMediaBundle

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst [hunk 19]
```diff
@@ -79,8 +79,9 @@ Block types:
  - ``sonata_mosaic_description``: this block will be always on screen and should represent the entity's name.
 
 
-The ``ObjectMetadata`` object is returned by the related admin class,
-for instance the MediaBundle defines the method as::
+The ``ObjectMetadata`` object is returned by the related admin class, and can be
+used to define which image field from the entity will be displayed if available.
+For instance, the SonataMediaBundle defines the method as::
```
Comment (OskarStark) [inline docs/cookbook/recipe_customizing_a_mosaic_list.rst]: Please use .. code-block:: php instead of ::

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst (hunk 18 above)
Comment (jo66) [reply docs/cookbook/recipe_customizing_a_mosaic_list.rst]: Done.

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst (hunk 19 above)
Comment (jo66) [reply docs/cookbook/recipe_customizing_a_mosaic_list.rst]: Done.

**Comment by jo66** [comment]
@greg0ire 
> When I register this bundle, there should be an alias pointing from `Sonata\MediaBundle\Provider\Pool` to `'@sonata.media.pool'`

Sorry but I don't understand, isn't it already defined in media bundle config/provider.xml:
```
        <!-- ... -->
        <parameter key="sonata.media.pool.class">Sonata\MediaBundle\Provider\Pool</parameter>
    </parameters>
    <services>
        <service id="sonata.media.pool" class="%sonata.media.pool.class%" public="true">
            <argument/>
        </service>
        <!-- ... -->
```
?

**Comment by greg0ire** [comment]
No it isn't. What you see here is:

- a parameter with key `sonata.media.pool.class` and `Sonata\MediaBundle\Provider\Pool` as a value
- a service with id `sonata.media.pool` and class `Sonata\MediaBundle\Provider\Pool`

However, no service alias from `Sonata\MediaBundle\Provider\Pool` to `'@sonata.media.pool'`.

Here is [the Symfony documentation about aliases](https://symfony.com/doc/current/service_container/alias_private.html#aliasing).
Here is [the best practice I would like you to implement](https://symfony.com/doc/4.1/bundles/best_practices.html#services).

> In addition, services not meant to be used by the application directly, should be defined as private. For public services, aliases should be created from the interface/class to the service id. For example, in MonologBundle, an alias is created from Psr\Log\LoggerInterface to logger so that the LoggerInterface type-hint can be used for autowiring.

Your mission, should you choose to accept it, is to open a PR on the media bundle creating the alias described above in XML (click the XML tab on the first link to see how it should look). Make sure to use `Sonata\MediaBundle\Provider\Pool` as id for the alias, not the other way around.

**Comment by jo66** [comment]
@greg0ire Mission accepted. Thanks a lot for explanations.
